File: testsuite/tests/ada/0017-callback-exceptions/cpp/main.cpp

#include <cstddef>
#include <cstdint>
#include <iostream>

#include "test.h"

void foo() {
  throw gnatpolyglot::ada::exceptions::ProgramError("bar");
}

int main() {
  try {
    test::get_callback()();
  } catch (const gnatpolyglot::ada::exceptions::ProgramError &e) {
    std::cout << "C++ caught : " << e.what() << "\n";
  }
  try {
    test::call(foo);
  } catch (const gnatpolyglot::ada::exceptions::ProgramError &e) {
    std::cout << "C++ caught : " << e.what() << "\n";
  }
}

File: testsuite/tests/ada/0017-callback-exceptions/java/Main.java
import com.adacore.libtest.test.TestPackage;

import com.adacore.gnatpolyglot.runtime.ada2java.ProgramError;
import com.adacore.gnatpolyglot.runtime.ada2java.PolyglotString;

public class Main {

    static void foo() {
        throw new ProgramError(new PolyglotString("bar"));
    }

    public static void main(String[] args) throws Throwable {
        try {
            TestPackage.getCallback().run();
        } catch (ProgramError e) {
            System.out.println("Java caught : " + e.getMessage());
        }

        try {
            TestPackage.call(Main::foo);
        } catch (ProgramError e) {
            System.out.println("Java caught : " + e.getMessage());
        }

    }
}
